onnx/checker.cc

```diff
@@ -676,39 +676,85 @@ void check_graph(const GraphProto& graph, const CheckerContext& ctx, const Lexic
   }
 }
 
+// Utilify function to get the imported version of domain from opset imports
+// Returns -1 if requested domain is not found in the opset_imports
+int get_version_for_domain(const std::string& domain, const std::unordered_map<std::string, int>& opset_imports) {
+  auto it = opset_imports.find(domain);
+  if (it == opset_imports.end()) {
+    return -1;
+  }
+
+  return it->second;
+}
+
+// Utility function to check compatibility of schema for 2 opsett versions for a given node.
+// Checks whether the schema for 2 versions is same. This is true when the opschema
+// does not change between versions.
+void check_opset_compatibility(
+    const NodeProto& node,
+    const CheckerContext& ctx,
+    const std::unordered_map<std::string, int>& func_opset_imports,
+    const std::unordered_map<std::string, int>& model_opset_imports) {
+  auto func_opset_version = get_version_for_domain(node.domain(), func_opset_imports);
+  auto model_opset_version = get_version_for_domain(node.domain(), model_opset_imports);
+
+  if (func_opset_version == -1) {
+    fail_check(
+        "No Opset registered for domain " + node.domain());
+  }
+
+  if (model_opset_version == -1) {
+    // model does not include opset import for a node present in function body.
+    // This is ok as along as the opset import is present in function level opset imports.
+    return;
+  }
+
+  const auto* schema_for_function_import =
+      ctx.get_schema_registry()->GetSchema(node.op_type(), func_opset_version, node.domain());
+
+  const auto* schema_for_model_import =
+      ctx.get_schema_registry()->GetSchema(node.op_type(), model_opset_version, node.domain());
+
+  if (schema_for_function_import && schema_for_model_import &&
+      schema_for_function_import->since_version() == schema_for_model_import->since_version()) {
+    // The since versions of schema for both opset imports match. This means they are compatible.
+    return;
+  }
+
+  fail_check(
+      "Opset import for domain " + node.domain() + " in function op " + node.op_type() +
+      "is not compatible with the version imported my model. FunctionOp imports version " +
+      ONNX_NAMESPACE::to_string(func_opset_version) + "whereas model imports version " +
+      ONNX_NAMESPACE::to_string(model_opset_version));
+}
+
 void check_function(const FunctionProto& function, const CheckerContext& ctx, const LexicalScopeContext& parent_lex) {
   enforce_non_empty_field(function, name);
-  enforce_has_field(function, since_version);
+  
+  if (ctx.get_ir_version() >= 0x00000008) {
+    enforce_has_field(function, domain);
```

Comment:
Is there a sensible/backwards-compatible default value for `domain` if it is unspecified?